src/Hadrian/Haskell/Cabal/Parse.hs

```diff
@@ -8,36 +8,287 @@
+                addFlag name       = C.insertFlagAssignment (C.mkFlagName name) True
+
+    let (Right (pd,_)) = C.finalizePackageDescription flags (const True) platform (C.compilerInfo compiler) [] gpd
+    -- depPkgs are all those packages that are needed. These should be found in
+    -- the known build packages.  Even if they are not build in this stage.
+    let depPkgs = map (findPackageByName' . C.unPackageName . C.depPkgName)
+                . flip C.enabledBuildDepends C.defaultComponentRequestedSpec $ pd
+          where findPackageByName' p = case findPackageByName p of
+                  Just p' -> p'
+                  Nothing -> error $ "Failed to find package: " ++ show p
+    return $ Cabal (C.unPackageName . C.pkgName . C.package $ pd)
+                   (C.display . C.pkgVersion . C.package $ pd)
+                   (C.synopsis pd)
+                   gpd
+                   pd
+                   depPkgs
+
+configurePackage :: Context -> Action ()
+configurePackage context@Context {..} = do
+    Just (Cabal _ _ _ gpd _pd depPkgs) <- readCabalFile context
+
+    -- Stage packages are those we have in this stage.
+    stagePkgs <- stagePackages stage
+    -- we'll need those package in our package database.
+    need =<< sequence [ pkgConfFile (context { package = pkg }) | pkg <- depPkgs, pkg `elem` stagePkgs ]
+
+    -- figure out what hooks we need.
+    hooks <- case C.buildType (C.flattenPackageDescription gpd) of
+          C.Configure -> pure Hooks.autoconfUserHooks
+          -- time has a "Custom" Setup.hs, but it's actually Configure
+          -- plus a "./Setup test" hook. However, Cabal is also
+          -- "Custom", but doesn't have a configure script.
+          C.Custom ->
+              do configureExists <- doesFileExist (replaceFileName (unsafePkgCabalFile package) "configure")
+                 if configureExists
+                     then pure Hooks.autoconfUserHooks
+                     else pure Hooks.simpleUserHooks
+          -- not quite right, but good enough for us:
+          _ | package == rts ->
+              -- don't try to do post conf validation for rts.
+              -- this will simply not work, due to the ld-options,
+              -- and the Stg.h.
+              pure $ Hooks.simpleUserHooks { Hooks.postConf = \_ _ _ _ -> return () }
+            | otherwise -> pure Hooks.simpleUserHooks
+
+
+    case pkgCabalFile package of
+      Nothing -> error "No a cabal package!"
+      Just _ -> do
+        -- compute the flaglist
+        flagList <- interpret (target context (CabalFlags stage) [] []) =<< args <$> flavour
+        -- compute the cabal conf args
+        argList <- interpret (target context (GhcCabal Conf stage) [] []) =<< args <$> flavour
+        liftIO $ do
+          Hooks.defaultMainWithHooksNoReadArgs hooks gpd (argList ++ ["--flags=" ++ unwords flagList])
+
+-- XXX: move this somewhere else. This is logic from ghc-cabal
```

Comment:
Is this function needed?

If we have `copyPackage` and `registerPackage`, do we really need `GhcCabal Copy` and `GhcCabal Reg`? If yes, what is the difference between them?

Reply:
It turns out we don't need `GhcCabal {Copy, Reg}`. Nice catch :)